### PR TITLE
2.x Extend interest subscription protocol to support snapshot subscriptions.

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannel.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+/**
+ * Interest channel for snapshot retrievals. Snapshots are one-time subscribe
+ * operations, that result in channel termination as soon as the last snapshot
+ * item is returned from a server.
+ *
+ * @author Tomasz Bak
+ */
+public interface SnapshotInterestChannel extends ServiceChannel {
+    Observable<InstanceInfo> forSnapshot(Interest<InstanceInfo> interest);
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannel.java]: This naming makes more sense. Can you also change the corresponding method name for InterestChannel from "change()" to "forInterest()" as well?

File: eureka-server/src/main/java/com/netflix/eureka2/server/channel/InterestChannelImpl.java [hunk 2]
```diff
@@ -53,9 +60,24 @@ public void call(Object message) {
                  * invocation), we do not need to worry about thread safety here and hence we can safely do a
                  * isRegistered -> register kind of actions without worrying about race conditions.
                  */
-                if (message instanceof InterestRegistration) {
+                if (message instanceof SnapshotRegistration) {
+                    switch (state.get()) {
+                        case Idle:
+                            sendSnapshot(((SnapshotRegistration) message).getInterests());
+                            state.set(STATES.Closed);
+                            metrics.stateTransition(STATES.Idle, STATES.Open);
```
Comment (qiangdavidliu) [inline eureka-server/src/main/java/com/netflix/eureka2/server/channel/InterestChannelImpl.java]: What is the state diagram for the snapshot request? The previous line sets the state to Closed but the metric transition only moves from Idle to Open.

File: eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                    public Observable<InstanceInfo> call(final MessageConnection connection) {
+                        return connection
+                                // Subscribe to interest set
+                                .submitWithAck(new SnapshotRegistration(interest))
+                                .cast(Object.class)
+                                .concatWith(connection.incoming())
+                                        // Handle change notification stream
+                                .takeWhile(END_OF_STREAM_FILTER)
+                                .filter(INVALID_MESSAGE_FILTER)
+                                .doOnNext(new Action1<Object>() {
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java]: You can merged the doOnNext() here into the map()

File: eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                                    @Override
+                                    public void call(Object message) {
+                                        connection.acknowledge().subscribe();
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java]: What happens if the ack here onErrors?

File: eureka-client/src/main/java/com/netflix/eureka2/client/resolver/EurekaServerResolver.java [hunk 5]
```diff
@@ -17,127 +17,153 @@
...
+            @Override
+            public Observable<Server> call(Set<Server> newServers) {
+                if (!newServers.isEmpty()) {
+                    updateLoadBalancer(newServers);
+                } else {
+                    logger.info("resolver returned empty server list; keeping the previous data");
+                }
+                return loadBalancer.choose();
+            }
+        }).onErrorResumeNext(new Func1<Throwable, Observable<Server>>() {
```
Comment (qiangdavidliu) [inline eureka-client/src/main/java/com/netflix/eureka2/client/resolver/EurekaServerResolver.java]: Shouldn't this onErrorResumeNext() be before the flatMap()?

File: eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannel.java (hunk 1 above)
Comment (tbak) [inline eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannel.java]: I would decouple ClientInterestChannel from InterestChannel (rename to ServerInterestChannel). We have two set of methods, and having common interface is artificial.

File: eureka-server/src/main/java/com/netflix/eureka2/server/channel/InterestChannelImpl.java (hunk 2 above)
Comment (tbak) [inline eureka-server/src/main/java/com/netflix/eureka2/server/channel/InterestChannelImpl.java]: I will fix this. We should transition to close.

File: eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java (hunk 4 above)
Comment (tbak) [inline eureka-client/src/main/java/com/netflix/eureka2/client/channel/SnapshotInterestChannelImpl.java]: Now we only log error messages if acks fail. I will add logging here.
In the future we will use acks for back pressure.

File: eureka-client/src/main/java/com/netflix/eureka2/client/resolver/EurekaServerResolver.java (hunk 5 above)
Comment (tbak) [inline eureka-client/src/main/java/com/netflix/eureka2/client/resolver/EurekaServerResolver.java]: It cannot be there.

**Pull request merged**